### ISSUE TITLE
Coroutine support - promise.get_future()?

### ISSUE THREAD

**Issue by tom-binary** [Naios/fluent.cpp#4]
First off, thank you for the library - the approach is a refreshing improvement over `std::future` and other options, and making it executor-agnostic was an excellent choice!

This is more of a question than an issue, but might lead to a doc/code patch eventually :smile: 

In a85040b0c2d2936ec65a81bdebdea586e90b7159 a basic skeleton for `std::experimental::coroutine_traits` was added. However, this currently does not appear to be functional - most of the code in there is commented out (and refers to `boost::promise` instead of `cti::continuable::promise`).

Is this under active development? The README implies that `co_await` is already usable:

https://github.com/Naios/continuable#coroutines

but aside from the exception codepath, that does not seem possible with the current 2.0.0 version? There also does not seem to be a way to get to the `continuable` from a `promise` instance (which I think would be required for `get_return_object` to work).

**Comment by Naios** [comment]
I'm glad that you like the library.

Yes you are absolutly right, currently it isn't possible to use continuables as return types from coroutines as shown below:

```c++
continuable<int> my_async_fn() {
  co_await ...;
  co_return 0;
}
```

Which makes it only possible to await on continuables where the return type of the async function is `void` or a type for which `coroutine_traits` is specialized.

Currently, this part isn't under active development since I didn't come to a satisfying solution for this and there are other parts such as waiting for an arbitrary amount of continuables (currently developed in the `devel` branch) which are way more important. I highly would appreciate a PR for this.

Yesterday I answered a similar question through mail to another user.
He tried to specialize a `handler_type` for continuables as described in [n3747](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3747.pdf) for boost asio ([boost::asio::use_future](http://www.boost.org/doc/libs/1_66_0/doc/html/boost_asio/reference/async_result.html)). This trait follows the same specilization principle as `coroutine_traits`.

Most of the customization traits of coroutines or the universal model handler_type are sadly
designed for eager futures which means we start the continuation chain directly.
Continuable uses lazy chaining and dispatches the chain once all continuations are known.
This makes it possible to stay zero cost.

Additionally, I'm hoping this answers your question about `get_return_object`:
- `std::future`: is designed in an eager way: we create a shared state and execute the action immediately, additionally we need space for attaching handlers through `then`. Thus the future and the promise are available instantly.
- `cti::continuable` is designed lazily: we build up a continuation chain in which the promise stores all callables which should be invoked when resolving it. Thus we can't retrieve the promise until the task is requested. This makes it really difficult to implement a `coroutine_traits` since we would need to insert a placeholder for an arbitrary promise. This is the trick I'm using to make continuables zero cost mostly when not applying any type erasure.

For coroutines, it should still be possible to start the coroutine and suspend it initially until we request its result. However, this won't be zero cost.

I think the standard facilities should be changed in order to support such a lazy execution model.

**Comment by tom-binary** [comment]
Okay, good to know - and thanks for the quick response.

I'll have a go at some test cases this week. That initially-suspended approach seems feasible: will raise a separate PR if I get anywhere!

**Comment by Naios** [comment]
@binarytom did you made any progress on this?

**Comment by Naios** [comment]
@binarytom I fixed this in be657109. Sadly this wasn't possible without any type erasure.
